File: bin/exam.py
#!/usr/bin/python3.1

#Copyright 2011 Newcastle University
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.


import re
from textile import textile
import examparser
import sys
import os

class ExamError(Exception):
	def __init__(self,message,hint=''):
		self.message = message
		self.hint = hint
	
	def __str__(self):
		msg = self.message
		if self.hint:
			msg += '\nPossible fix: '+self.hint
		return msg

#data is a DATA object. attr is either a single variable name or a list of names. obj is the object to load the data into. altname is the name of the object property to fill, if different from attr
#if attr is in data, then obj.attr = data[attr], otherwise no change
def tryLoad(data,attr,obj,altname=''):
	if type(attr)==list:
		for x in attr:
			tryLoad(data,x,obj)
		return
	else:
		if not altname:
			altname = attr
		attr = attr.lower()
		if attr in data:
			if type(obj)==dict:
				obj[altname]=data[attr]
			else:
				setattr(obj,altname,data[attr])

#convert a textile block of content into html, wrapped in a <content> tag and optionally an <html> tag too.
def makeContentNode(s,doTextile=True):
	s=str(s)
	s='\n'.join([x.lstrip() for x in s.split('\n')])	#textile doesn't like whitespace at the start of a line, but I do
	if doTextile:
		s=textile(s)
	else:
		s = re.sub(r'&(?!#?\w+;)','&amp;',s)
		s='<span>'+s+'</span>'
	return etree.fromstring('<content>'+s+'</content>')

#exam object
class Exam:
	name = ''						#title of exam
	duration = 0					#allowed time for exam, in seconds
	percentPass = 0					#percentage classified as a pass
	shuffleQuestions = False		#randomise question order?
	showactualmark = True			#show student's score to student?
	showtotalmark = True			#show total marks available to student?
	showanswerstate = True			#show right/wrong on questions?
	allowrevealanswer = True		#allow student to reveal answer to question?
	adviceType = 'onreveal'			#when is advice shown? 'onreveal' only option at the moment
	adviceThreshold = 0		#reveal advice if student scores less than this percentage

	def __init__(self,name='Untitled Exam'):
		self.name = name
		self.navigation = {	
				'allowregen': False,				#allow student to re-randomise a question?
				'reverse': True,
				'browse': True,
				'showfrontpage': True,
				'onadvance': Event('onadvance','none','You have not finished the current question'),
				'onreverse': Event('onreverse','none','You have not finished the current question'),
				'onmove': Event('onmove','none','You have not finished the current question')
			}

		self.timing = { 
				'timeout': Event('timeout','none',''),
				'timedwarning': Event('timedwarning','none','')
			}

		self.rulesets = {}

		self.functions = []
		self.variables = []
		
		self.questions = []

		self.resources = []
		self.extensions = []
	
	@staticmethod
	def fromstring(string):
		try:
			data = examparser.ExamParser().parse(string)
			exam = Exam.fromDATA(data)
			return exam
		except examparser.ParseError as err:
			print('Parse error: ',str(err))
			raise

	@staticmethod
	def fromDATA(data):
		exam = Exam()
		tryLoad(data,['name','duration','percentPass','shuffleQuestions','resources','extensions'],exam)

		if 'navigation' in data:
			nav = data['navigation']
			tryLoad(nav,['allowregen','reverse','browse','showfrontpage'],exam.navigation)
			for event in ['onadvance','onreverse','onmove']:
				if event in nav:
					tryLoad(nav[event],['action','message'],exam.navigation[event])

		if 'timing' in data:
			timing = data['timing']
			for event in ['timeout','timedwarning']:
				if event in timing:
					tryLoad(timing[event],['action','message'],exam.timing[event])

		if 'feedback' in data:
			tryLoad(data['feedback'],['showactualmark','showtotalmark','showanswerstate','allowrevealanswer'],exam)
			if 'advice' in data['feedback']:
				advice = data['feedback']['advice']
				tryLoad(advice,'type',exam,'adviceType')
				tryLoad(advice,'threshold',exam,'adviceThreshold')

		if 'rulesets' in data:
			rulesets = data['rulesets']
			for name in rulesets.keys():
				l=[]
				for rule in rulesets[name]:
					if isinstance(rule,str):
						l.append(rule)
					else:
						l.append(SimplificationRule.fromDATA(rule))
				exam.rulesets[name] = l

		if 'functions' in data:
			functions = data['functions']
			for function in functions.keys():
				exam.functions.append(Function.fromDATA(function,functions[function]))

		if 'variables' in data:
			variables = data['variables']
			for variable in variables.keys():
				exam.variables.append(Variable(variable,variables[variable]))

		if 'questions' in data:
			for question in data['questions']:
				exam.questions.append(Question.fromDATA(question))

		return exam


	#return a simplified object suitable for putting into JSON/.exam format
	def export(self):
		obj = {
			'name': self.name,
			'percentPass': self.percentPass,
			'shuffleQuestions': self.shuffleQuestions,
			'duration': self.duration,
			'settings': {
				'navigation': {
					'allowregen': self.navigation['allowregen'],
					'reverse': self.navigation['reverse'],
					'browse': self.navigation['browse'],
					'showFrontPage': self.navigation['showfrontpage'],
					'events': {
						'advance': self.navigation['onadvance'].export(),
						'reverse': self.navigation['onreverse'].export(),
						'move': self.navigation['onmove'].export()
					}
				},
				'timing': {
					'timeout': self.timing['timeout'].export(),
					'timedwarning': self.timing['timedwarning'].export()
				},
				'feedback': {
					'showActualMark': self.showactualmark,
					'showTotalMark': self.showtotalmark,
					'showAnswerState': self.showanswerstate,
					'allowRevealAnswer': self.allowrevealanswer,
				},
				'rulesets': { name: [rule.export() if isinstance(rule,SimplificationRule) else rule for rule in rules] for name,rules in self.rulesets.items() }
			},
			'functions': { function.name: function.export() for function in self.functions },
			'variables': { variable.name: variable.definition for variable in self.variables },
			'questions': [ q.export() for q in self.questions ]
		}
		return obj

class SimplificationRule:
	pattern = ''
	result = ''

	def __init__(self):
		self.conditions = []

	@staticmethod
	def fromDATA(data):
		rule=SimplificationRule()
		tryLoad(data,['pattern','conditions','result'],rule)
		return rule

	def export(self):
		return {
			'pattern': self.pattern,
			'result': self.result,
			'conditions': self.conditions
		}


class Event:
	kind = ''
	action = 'none'
	message = ''

	def __init__(self,kind,action,message):
		self.kind = kind
		self.action = action
		self.message = message

	def export(self):
		return {
			'action': self.action,
			'message': self.message
		}

class Question:
	name = 'Untitled Question'
	statement =''
	advice = ''

	def __init__(self,name='Untitled Question'):
		self.name = name

		self.parts = []
		self.variables = []
		self.functions = []

	@staticmethod
	def fromDATA(data):
		question = Question()
		tryLoad(data,['name','statement','advice'],question)

		if 'parts' in data:
			parts = data['parts']
			for part in parts:
				question.parts.append(Part.fromDATA(part))

		if 'variables' in data:
			variables = data['variables']
			for variable in variables.keys():
				question.variables.append(Variable(variable,variables[variable]))
		
		if 'functions' in data:
			functions = data['functions']
			for function in functions.keys():
				question.functions.append(Function.fromDATA(function,functions[function]))

		return question

	def export(self):
		return {
			'name': self.name,
			'statement': self.statement,
			'advice': self.advice,
			'parts': [part.export() for part in self.parts],
			'variables': { variable.name: variable.definition for variable in self.variables },
			'functions': { function.name: function.export() for function in self.functions }
		}

class Variable:
	name = ''
	definition = ''

	def __init__(self,name,definition):
		self.name = name
		self.definition = definition
	
class Function:
	name = ''
	type = ''
	definition = ''

	def __init__(self,name):
		self.name = name
		self.parameters = {}
	
	@staticmethod
	def fromDATA(name,data):
		function = Function(name)
		tryLoad(data,['parameters','type','definition'],function)
		return function
	
	def export(self):
		return {
			'type': self.type,
			'definition': self.definition,
			'parameters': [{name: x[0], type: x[1]} for x in self.parameters]
		}

class Part:
	prompt = ''
	kind = ''
	stepsPenalty = 0
	enableMinimumMarks = True
	minimumMarks = 0

	def __init__(self,marks,prompt=''):
		self.marks = marks
		self.prompt = prompt
		self.steps = []

	@staticmethod
	def fromDATA(data):
		kind = data['type'].lower()
		partConstructors = {
				'jme': JMEPart,
				'numberentry': NumberEntryPart,
				'patternmatch': PatternMatchPart,
				'1_n_2': MultipleChoicePart,
				'm_n_2': MultipleChoicePart,
				'm_n_x': MultipleChoicePart,
				'gapfill': GapFillPart,
				'information': InformationPart
			}
		if not kind in partConstructors:
			raise ExamError(
				'Invalid part type '+kind,
				'Valid part types are '+', '.join(sorted([x for x in partConstructors]))
			)
		part = partConstructors[kind].fromDATA(data)

		tryLoad(data,['stepsPenalty','minimumMarks','enableMinimumMarks'],part);

		if 'marks' in data:
			part.marks = data['marks']

		if 'prompt' in data:
			part.prompt = data['prompt']

		if 'steps' in data:
			steps = data['steps']
			for step in steps:
				part.steps.append(Part.fromDATA(step))

		return part
	
	def export(self):
		return {
			'type': self.kind,
			'marks': self.marks,
			'prompt': self.prompt,
			'stepspenalty': self.stepsPenalty,
			'enableminimummarks': self.enableMinimumMarks,
			'minimummarks': self.minimumMarks,
			'steps': [step.export() for step in self.steps]
		}


class JMEPart(Part):
	kind = 'jme'
	answer = ''
	answerSimplification = 'unitFactor,unitPower,unitDenominator,zeroFactor,zeroTerm,zeroPower,collectNumbers,zeroBase,constantsFirst,sqrtProduct,sqrtDivision,sqrtSquare,otherNumbers'
	checkingType = 'RelDiff'
	checkingAccuracy = 0		#real default value depends on checkingtype - 0.0001 for difference ones, 5 for no. of digits ones
	failureRate = 1
	vsetRangeStart = 0
	vsetRangeEnd = 1
	vsetRangePoints = 5

	def __init__(self,marks=0,prompt=''):
		Part.__init__(self,marks,prompt)

		self.maxLength = Restriction('maxlength',0,'Your answer is too long.')
		self.maxLength.length = 0
		self.minLength = Restriction('minlength',0,'Your answer is too short.')
		self.minLength.length = 0
		self.mustHave = Restriction('musthave',0,'Your answer does not contain all required elements.')
		self.notAllowed = Restriction('notallowed',0,'Your answer contains elements which are not allowed.')
	
	@staticmethod
	def fromDATA(data):
		part = JMEPart()
		tryLoad(data,['answer','answerSimplification','checkingType','failureRate','vsetRangePoints'],part)

		#default checking accuracies
		if part.checkingType.lower() == 'reldiff' or part.checkingType.lower() == 'absdiff':
			part.checkingAccuracy = 0.0001
		else:	#dp or sigfig
			part.checkingAccuracy = 5
		#get checking accuracy from data, if defined
		tryLoad(data,'checkingAccuracy',part)

		if 'maxlength' in data:
			part.maxLength = Restriction.fromDATA('maxlength',data['maxlength'],part.maxLength)
		if 'minlength' in data:
			part.minLength = Restriction.fromDATA('minlength',data['minlength'],part.minLength)
		if 'musthave' in data:
			part.mustHave = Restriction.fromDATA('musthave',data['musthave'],part.mustHave)
		if 'notallowed' in data:
			part.notAllowed = Restriction.fromDATA('notallowed',data['notallowed'],part.notAllowed)

		if 'vsetrange' in data and len(data['vsetrange']) == 2:
			part.vsetRangeStart = data['vsetrange'][0]
			part.vsetRangeEnd = data['vsetrange'][1]

		return part

	def export(self):
		obj = super(JMEPart,self).export()
		obj.update({
			'answer': {
				'correctAnswer': self.answer,
				'answerSimplification': self.answerSimplification,
				'checking': {
					'checkingType': self.checkingType,
					'checkingAccuracy': self.checkingAccuracy,
					'failureRate': self.failureRate,
					'range': {
						'start': self.vsetRangeStart,
						'end': self.vsetRangeEnd,
						'points': self.vsetRangePoints
					},
				},
				'maxLength': self.maxLength.export(),
				'minLength': self.minLength.export(),
				'mustHave': self.mustHave.export(),
				'notAllowed': self.notAllowed.export()
			}
		})
		return obj

class Restriction:
	message = ''
	length = -1
	showStrings = False

	def __init__(self,name='',partialCredit=0,message=''):
		self.name = name
		self.strings = []
		self.partialCredit = partialCredit
		self.message = message
	
	@staticmethod
	def fromDATA(name,data,restriction=None):
		if restriction==None:
			restriction = Restriction(name)
		tryLoad(data,['showStrings','partialCredit','message','length'],restriction)
		if 'strings' in data:
			for string in data['strings']:
				restriction.strings.append(string)

		return restriction

	def export(self):
		obj = {
			'message': self.message,
			'partialcredit': self.partialCredit
		}

		if self.length>=0:
			obj['length']=self.length

		if len(self.strings)>0:
			obj['strings'] = self.strings
			obj['showStrings'] = self.showStrings

		return obj

class PatternMatchPart(Part):
	kind = 'patternmatch'
	caseSensitive = False
	partialCredit = 0
	answer = ''
	displayAnswer = ''

	def __init__(self,marks=0,prompt=''):
		Part.__init__(self,marks,prompt)

	@staticmethod
	def fromDATA(data):
		part = PatternMatchPart()
		tryLoad(data,['caseSensitive','partialCredit','answer','displayAnswer'],part)

		return part

	def export(self):
		obj = super(PatternMatchPart,self).export()
		obj.update({
			'displayAnswer': self.displayAnswer,
			'correctAnswer': self.answer,
			'caseSensitive': self.caseSensitive,
			'partialCredit': self.partialCredit
		})
		return obj

class NumberEntryPart(Part):
	kind = 'numberentry'
	integerAnswer = False
	partialCredit = 0
	minvalue = 0
	maxvalue = 0
	inputStep = 1

	def __init__(self,marks=0,prompt=''):
		Part.__init__(self,marks,prompt)
	
	@staticmethod
	def fromDATA(data):
		part = NumberEntryPart()
		tryLoad(data,['integerAnswer','partialCredit','minvalue','maxvalue','inputStep'],part)
		if 'answer' in data:
			part.maxvalue = part.minvalue = data['answer']

		return part

	def export(self):
		obj = super(NumberEntryPart,self).export()
		obj.update({
			'answer': {
				'minValue': self.minvalue,
				'maxValue': self.maxvalue,
				'inputStep': self.inputStep,
				'integerAnswer': self.integerAnswer,
				'partialCredit': self.partialCredit
			}
		})
		return obj

class MultipleChoicePart(Part):
	minMarksEnabled = False
	minMarks = 0
	maxMarksEnabled = False
	maxMarks = 0
	minAnswers = 0
	maxAnswers = 0
	shuffleChoices = False
	shuffleAnswers = False
	displayType = 'radiogroup'
	displayColumns = 1
	warningType = 'warn'
	warningMessage = 'You have not selected the correct number of choices.'
	
	def __init__(self,kind,marks=0,prompt=''):
		self.kind = kind
		Part.__init__(self,marks,prompt)

		self.choices = []
		self.answers = []
		self.matrix = []

		self.distractors = []

	@staticmethod
	def fromDATA(data):
		kind = data['type']
		part = MultipleChoicePart(kind)
		displayTypes = {
				'1_n_2': 'radiogroup',
				'm_n_2': 'checkbox',
				'm_n_x': 'radiogroup'
		}

		part.displayType = displayTypes[kind]
		tryLoad(data,['minMarks','maxMarks','minAnswers','maxAnswers','shuffleChoices','shuffleAnswers','displayType','displayColumns','warningType','warningMessage'],part)

		if 'choices' in data:
			for choice in data['choices']:
				part.choices.append(choice)

		if 'answers' in data:
			for answer in data['answers']:
				part.answers.append(answer)
	
		if 'matrix' in data:
			part.matrix = data['matrix']
			if not isinstance(part.matrix[0],list):	#so you can give just one row without wrapping it in another array
				part.matrix = [[x] for x in part.matrix]

		if 'distractors' in data:
			part.distractors = data['distractors']
			if not isinstance(part.distractors[0],list):
				part.distractors = [[x] for x in part.distractors]

		return part

	def export(self):
		obj = super(MultipleChoicePart,self).export()
		obj.update({
			'minMarks': self.minMarks,
			'maxMarks': self.maxMarks,
			'minAnswers': self.minAnswers,
			'maxAnswers': self.maxAnswers,
			'shuffleChoices': self.shuffleChoices,
			'shuffleAnswers': self.shuffleAnswers,
			'displayType': self.displayType,
			'displayColumns': self.displayColumns,
			'warningType': self.warningType,
			'warningMessage': self.warningMessage,
			'choices': self.choices,
			'answers': self.answers,
			'matrix': self.matrix,
			'distractors': self.distractors
		})
		return obj

class InformationPart(Part):
	kind = 'information'

	def __init__(self,prompt=''):
		Part.__init__(self,0,prompt)
	
	@staticmethod
	def fromDATA(data):
		return InformationPart()

class GapFillPart(Part):
	kind = 'gapfill'

	def __init__(self,prompt=''):
		Part.__init__(self,0,prompt)
		
		self.gaps = []

	@staticmethod
	def fromDATA(data):
		part = GapFillPart()

		if 'gaps' in data:
			gaps = data['gaps']
			for gap in gaps:
				part.gaps.append(Part.fromDATA(gap))

		return part
	
	def export(self):
		obj = super(GapFillPart,self).export()
		obj.update({
			'gaps': [gap.export() for gap in self.gaps]
		})
		return obj

File: runtime/scripts/exam.js
/*
Copyright 2011 Newcastle University

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*/


Numbas.queueScript('scripts/exam.js',['json','timing','util','display','schedule','scorm-storage','pretendlms','math','question','jme-variables','jme-display'],function() {


// exam object keeps track of all info we need to know while exam is running
var Exam = Numbas.Exam = function()
{
	var parseBool = Numbas.util.parseBool;

	var json = this.json = Numbas.raw.examJSON;

	var exam = this;

	var tryLoad = function(path,to){
		to = to || exam;
		return Numbas.json.tryLoad(path,to,json);
	};

	['name','duration','percentPass','allQuestions','selectQuestions','shuffleQuestions'].map(tryLoad());

	this.navigation = Numbas.util.copyobj(this.navigation);
	['reverse','browse','showFrontPage'].map(tryLoad('settings.navigation',this.navigation));

	//get navigation events and actions
	var navEvents = this.navigation.events = {};

	['advance','reverse','move'].map(tryLoad('settings.navigation.events',navEvents));
	for( var ev in navEvents )
	{
		navEvents[ev] = new ExamEvent(navEvents[ev]);
	}

						
	//get timing events
	var timerEvents = this.timerEvents = {};
	['timeout','timedwarning'].map(tryLoad('settings.timing',timerEvents));
	for( var ev in timerEvents )
	{
		timerEvents[ev] = new ExamEvent(timerEvents[ev]);
	}
		
	//feedback
	['showActualMark','showTotalMark','showAnswerState','allowRevealAnswer','adviceType','adviceThreshold'].map(tryLoad('settings.feedback'));

	//rulesets
	this.rulesets = Numbas.util.copyobj(Numbas.jme.display.simplificationRules);
	for(var name in json.settings.rulesets)
	{
		var set = json.settings.rulesets[name].map(function(r) {
			if(typeof(r)=='string')
				return r;
			else
				return new Numbas.jme.display.Rule(r.pattern,r.conditions,r.result);
		});
		
		this.rulesets[name] = Numbas.jme.display.collectRuleset(set,this.rulesets);
	}

	this.rulesets['default'] = Numbas.jme.display.collectRuleset(['unitFactor','unitPower','unitDenominator','zeroFactor','zeroTerm','zeroPower','collectNumbers','zeroBase','constantsFirst','sqrtProduct','sqrtDivision','sqrtSquare','otherNumbers'],this.rulesets);

	//get text representation of exam duration
	this.displayDuration = this.duration>0 ? Numbas.timing.secsToDisplayTime( this.duration, true ) : '';

	this.totalQuestions = json.questions.length;

	//initialise display
	this.display = new Numbas.display.ExamDisplay(this);

}
Exam.prototype = {

	json: undefined,			//JSON object defining exam

	mode: 'entry',				//can be 	"entry" - exam not started yet
								//			"in progress" - exam started, not finished
                                //			"review" - looking at completed exam
                                //			"suspend" - exam is paused

	//exam properties
	name: '',					//title of exam
	mark: 0,					//total marks available in exam
	score: 0,					//student's current score
	percentPass: 0,				//percentage student must achieve to pass
	percentScore: 0,			//student's score as a percentage
	passed: false,				//did student pass the exam?

	//variables and functions
	functions: {},				//functions available to every question
	variables: {},				//variables available to every question
	
	//question selection
	totalQuestions: 0,			//how many questions are available?
	allQuestions: true,			//use all questions?
	selectQuestions: 0,			//how many questions to select, if not using all?
	shuffleQuestions: false,	//should the questions be shuffled?
	currentQuestion: undefined,	//current question object
	
	numQuestions: 0,			//number of questions in this sitting
	questionSubset: [],			//which questions from the pool to use? for reconstructing question list on resume
	questionList: [],			//Question objects, in order student will see them
		
	//navigation
	navigation: {
		showFrontPage: true,	//show the front page, or go straight to first question?
		reverse: false,			//can student navigate to previous question?
		browse: false,			//can student jump to any question they like?
		events: {}
	},

	timerEvents: {},			//events based on timing
	
	//timing
	duration: 0,				//how long is exam?
	displayDuration: '',//exam duration in h:m:s format
	timeoutAction: 'none',		//what to do when timer runs out
	timedWarningAction: 'none',	//warning 5 minutes before end?
	stopwatch: undefined,		//stopwatch object - updates timer every second
	endTime: undefined,			//time that the exam should stop
	timeRemaining: 0,			//seconds until end of exam
	timeSpent: 0,				//seconds exam has been in progress
	inProgress: false,			//is the exam in progress? False before starting, when paused, and after ending.

	start: Date(),				//time exam started
	stop: Date(),				//time exam finished
	
	//feedback
	showActualMark: false,		//show current score?
	showTotalMark: false,		//show total marks in exam?
	showAnswerState: false,		//tell student if answer is correct/wrong/partial ?
	allowRevealAnswer: false,	//allow 'reveal answer' button ?
	adviceType: '',				//something to do with when advice can be shown ??
	adviceThreshold: 0, 	//if student scores lower than this percentage on a question, the advice is displayed

	display: undefined,			//display code

	//stuff to do when starting exam afresh
	init: function()
	{
		var job = Numbas.schedule.add;
		var exam = this;
		job(function() {
			exam.functions = Numbas.jme.variables.makeFunctions(exam.json.functions);
			exam.variables = Numbas.jme.variables.makeVariables(exam.json.variables,exam.functions);
		});
		job(exam.chooseQuestionSubset,exam);			//choose questions to use
		job(exam.makeQuestionList,exam);				//create question objects
		job(Numbas.store.init,Numbas.store,exam);		//initialise storage
	},

	//restore previously started exam from storage
	load: function()
	{
		var suspendData = Numbas.store.load();	//get saved info from storage

		this.timeRemaining = suspendData.timeRemaining;
		this.questionSubset = suspendData.questionSubset;
		this.start = new Date(suspendData.start);
		this.score = suspendData.score;

		this.makeQuestionList(true);

		if(suspendData.location!==undefined)
			this.changeQuestion(suspendData.location);

		for(i=0; i<this.numQuestions; i++)
		{
			if(this.questionList[i].visited)
				this.questionList[i].display.showScore();
		}
		this.display.showScore();
	},


	//xmlize for info pages and so on
	xmlize: function()
	{
		var obj = {};
		var dontwant = ['xml','questionList','stopwatch','display','currentQuestion','navigationEvents','rulesets','functions','variables'];
		for( var x in this )
		{
			if(!(dontwant.contains(x) || typeof(this[x])=='function'))
				obj[x]=this[x];
		}

		return Sarissa.xmlize(obj,'exam');
	},

	//decide which questions to use and in what order
	chooseQuestionSubset: function()
	{
		//get all questions out of XML
		var tmpQuestionList = new Array();

		//decide how many questions in this sitting
		if( this.allQuestions )
		{
			this.numQuestions = this.totalQuestions;
		}
		else
		{
			this.numQuestions = Math.min(this.totalQuestions,this.selectQuestions);
		}

		//shuffle questions?
		this.questionSubset = [];
		if(this.shuffleQuestions)
		{
			this.questionSubset=Numbas.math.deal(this.numQuestions);
		}
		else	//otherwise just pick required number of questions from beginning of list
		{
			this.questionSubset = Numbas.math.range(this.numQuestions);
		}

		if(this.questionSubset.length==0)
		{
			Numbas.display.showAlert("This exam contains no questions! Check the .exam file for errors.");
		}
	},

	//having chosen which questions to use, make question list and create question objects
	//if loading, need to restore randomised variables instead of generating anew
	makeQuestionList: function(loading)
	{
		var job = Numbas.schedule.add;

		this.questionList = [];
		
		var questions = this.json.questions;
		for(var i = 0; i<this.questionSubset.length; i++) 
		{
			job(function(i)
			{
				var question = new Numbas.Question( questions[this.questionSubset[i]], i, loading, this.variables, this.functions );
				this.questionList.push(question);
			},this,i);
		}

		job(function() {
			//calculate max marks available in exam
			this.mark = 0;

			//go through the questions and recalculate the part scores, then the question scores, then the exam score
			for( i=0; i<this.numQuestions; i++ )
			{
				this.mark += this.questionList[i].marks;
			}
		},this);
	},
	
	//begin exam
	begin: function()
	{
		this.start = new Date();        //make a note of when the exam was started
		this.endTime = new Date(this.start.getTime()+this.duration*1000);	//work out when the exam should end

		this.changeQuestion(0);			//start at the first question!

		this.updateScore();				//initialise score

		//set countdown going
		if(this.mode!='review')
			this.startTiming();

		this.display.showQuestion();	//display the current question

	},

	pause: function()
	{
		this.endTiming();
		this.display.showInfoPage('suspend');

		Numbas.store.pause();
	},

	resume: function()
	{
		this.startTiming();
		this.display.showQuestion();
	},

	//set the stopwatch going
	startTiming: function()
	{
		this.inProgress = true;
		this.stopwatch = {
			start: new Date(),
			end: new Date((new Date()).getTime() + this.timeRemaining*1000),
			oldTimeSpent: this.timeSpent,
			id: setInterval(function(){exam.countDown();}, 1000)
		};

		if( this.duration > 0 )
			this.display.showTiming();
			
		else
			this.display.hideTiming();

		var exam = this;
		this.countDown();
	},

	//display time remaining and end exam when timer reaches zero
	countDown: function()
	{
		var t = new Date();
		this.timeSpent = this.stopwatch.oldTimeSpent + (t - this.stopwatch.start)/1000;

		if(this.duration > 0)
		{
			this.timeRemaining = Math.ceil((this.endTime - t)/1000);
			this.display.showTiming();

			if(this.duration > 300 && this.timeRemaining<300 && !this.showedTimeWarning)
			{
				this.showedTimeWarning = true;
				var e = this.timerEvents['timedwarning'];
				if(e && e.action=='warn')
				{
					Numbas.display.showAlert(e.message);
				}
			}
			else if(this.timeRemaining===0)
			{
				var e = this.timerEvents['timeout'];
				if(e && e.action=='warn')
				{
					Numbas.display.showAlert(e.message);
				}
				this.end();
			}	
		}
	},

	//stop the stopwatch
	endTiming: function()
	{
		this.inProgress = false;
		clearInterval( this.stopwatch.id );
	},


	//recalculate and display student's total score
	updateScore: function()
	{
		this.calculateScore();
		this.display.showScore();	
	},

	calculateScore: function()
	{
		this.score=0;
		for(var i=0; i<this.questionList.length; i++)
			this.score += this.questionList[i].score;
	},

	//call this when student wants to move between questions
	//will check move is allowed and if so change question and update display
	tryChangeQuestion: function(i)
	{
		if(i<0 || i>=this.numQuestions)
			return;

		var currentQuestion = this.currentQuestion;
		if(!currentQuestion)
			return;

		if(i==currentQuestion.number)
			return;

		var exam = this;
		function go()
		{
			exam.changeQuestion(i);
			exam.display.showQuestion();
		}

		if(currentQuestion.answered)
		{
			go();
		}
		else
		{
			//work out what action is happening
			var event = '';
			if(i==currentQuestion.number-1)		//reversing
				event='reverse';
			else if(i==currentQuestion.number+1)	//advancing
				event='advance';
			else								//jumping
				event='move';

			var eventObj = this.navigation.events[event];
			switch( eventObj.action )
			{
			case 'none':
				go();
				break;
			case 'warnifunattempted':
				Numbas.display.showConfirm(eventObj.message+'<p>Proceed anyway?</p>',go);
				break;
			case 'preventifunattempted':
				Numbas.display.showAlert(eventObj.message);
				break;
			}
		}
	},

	//actually change the current question
	changeQuestion: function(i)
	{
		this.currentQuestion = this.questionList[i];
		if(!this.currentQuestion)
		{
			throw(new Error("This exam contains no questions! Check the .exam file for errors."));
		}
		this.currentQuestion.visited = true;
		Numbas.store.changeQuestion(this.currentQuestion);
	},

	end: function()
	{
		//get time of finish
		this.stop = new Date();
		
		//stop the stopwatch
		this.endTiming();

		//work out summary info
		this.percentScore = Math.round(100*this.score/this.mark);
		this.passed = this.percentScore >= this.percentPass;

		//construct report object
		var report = this.report = 
		{	examsummary: {	name: this.name,
							numberofquestions: this.numQuestions, 
							mark: this.mark,
							passpercentage: this.percentPass,
							duration: this.displayDuration 
						 },
			performancesummary: {	start: this.start.toGMTString(),
									stop: this.stop.toGMTString(),
									timespent: Numbas.timing.secsToDisplayTime(this.timeSpent),
									score: this.score,
									percentagescore: this.percentScore,
									passed: this.passed,
									result: (this.passed ? 'Passed' :'Failed')
								},
			questions: new Array()
		};

		//construct reports for each question
		var examQuestionsAttempted = 0;
		for(var j=0; j<this.questionList.length; j++)
		{
			var question = this.questionList[j];
			if(question.answered)
				examQuestionsAttempted++;

			report.questions.push({question: {	number: question.number+1,
												name: question.name,
												marks: question.marks,
												score: question.score } });
		}
		report.performancesummary.questionsattempted = examQuestionsAttempted;


		//send result to LMS, and tell it we're finished
		Numbas.store.end();

		//display the results
		this.display.showInfoPage( 'result' );
	},

	exit: function()
	{
		this.display.showInfoPage('exit');
	}
};

function ExamEvent(json)
{
	['type','action','message'].map(Numbas.json.tryLoad('',this,json));
	this.message = textile(this.message);
}
ExamEvent.prototype = {
	type: '',
	action: 'none',
	message: ''
};

});
